### PR TITLE
Implement compressing\decompressing using zlib while reading\writing cnv gridfs stream

### PR DESCRIPTION
**Pull request by oleksiivorobiov**

### CODE REVIEW

File: src/mongoc/mongoc-stream-cnv-gridfs.c [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+   mongoc_stream_cnv_gridfs_t *stream;
+   mongoc_stream_t *gridfs_stream;
+
+   gridfs_stream = mongoc_stream_gridfs_new (file);
+   if (!gridfs_stream)
+      return NULL;
+
+   stream = bson_malloc0 (sizeof *stream);
+   stream->stream.type = MONGOC_STREAM_GRIDFS;
+   stream->stream.destroy = _mongoc_stream_cnv_gridfs_destroy;
```
Comment (jsbattig) [inline src/mongoc/mongoc-stream-cnv-gridfs.c:238]: @alexeyvo, so the way mongo people implemented this methods is doing the calls via some kind of virtual method table that you can hook into?

File: src/mongoc/mongoc-stream-file.c [hunk 2]
```diff
@@ -197,7 +197,7 @@ mongoc_stream_file_new_for_path (const char *path,  /* IN */
    bson_return_val_if_fail (path, NULL);
 
 #ifdef _WIN32
-   if (_sopen_s (&fd, path, (flags | _O_BINARY), _SH_DENYNO, 0) != 0) {
+   if (_sopen_s (&fd, path, (flags | _O_BINARY), _SH_DENYNO, mode) != 0) {
```
Comment (jsbattig) [inline src/mongoc/mongoc-stream-file.c:200]: @alexeyvo , is this a bug fix?

File: src/mongoc/mongoc-stream-cnv-gridfs.c (hunk 1 above)
Comment (oleksiivorobiov) [inline src/mongoc/mongoc-stream-cnv-gridfs.c:238]: take look on this structure:

```
typedef struct
{
   mongoc_stream_t       stream;
   mongoc_stream_t      *gridfs_stream;
   ...
} mongoc_stream_cnv_gridfs_t;
```

first field is mongoc_stream_t, so when mongoc_stream_cnv_gridfs_t casted to mongoc_stream_t addressess of  methods read\write\etc of this first field will be called. That's some kind of inheriting\implementing interface. We create stream using mongoc_stream_gridfs_new, but it returns mongoc_stream_t, not mongoc_stream_cnv_gridfs_t, so we just call "interface method" mongoc_stream_read() passing mongoc_stream_t.

File: src/mongoc/mongoc-stream-file.c (hunk 2 above)
Comment (oleksiivorobiov) [inline src/mongoc/mongoc-stream-file.c:200]: Yes. I submited few bugfixes(including this) to mongodb but they didn't answer yet.

**Pull request merged**